### PR TITLE
[python-package] [docs] Expand class docs for Dataset

### LINKED ISSUE #6511: [python-package] Training data partially available when loading exported lgb.Dataset
## Summary

From a binary file exported using the `lgb.Dataset.save_binary()` method, it is possible to retrieve the feature names (list of string) and the labels (y, as a numpy array) 
It is not however possible to retrieve the data, as a numpy array.
Both the `lgb.Dataset.data` and `lgb.Dataset.get_data()` will send back a string with the name of the binary file.
It should, in my opinion, send back the numpy array, just as the `label`.
Note it seems as well that the `group` is also not accessible.

The data is effectively contained in the binary file, because we are able to load and train a booster with it.
We should be able to get it from the newly created lgb.Dataset object

## Motivation

This would allow us to properly investigate the exported datasets.
Like computing some statistics (mean, stdv etc) to understand our dataset, before using it again (or not) to train a booster.

## Description

The `.data` atttribute and/or `.get_data()` method send back the proper numpy array.
The `group` and/or `.get_group()` send back the proper list of int

## References
I executed the following script on the [MQ2008](https://onedrive.live.com/?authkey=%21ACnoZZSZVfHPJd0&cid=8FEADC23D838BDA8&id=8FEADC23D838BDA8%21114&parId=8FEADC23D838BDA8%21107&o=OneUp), a microsoft open ranking dataset
You should be able to reproduce the same results.
I also provide an output so that you can understand what are the atttibutes and method that should have a modified behaviour

```python
import logging

import lightgbm as lgb
from sklearn.datasets import load_svmlight_file

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s : %(message)s",
)
logging.info(f"lgb.__version__ = {lgb.__version__}")


# Function to load group information
def load_group_info(file_path):
    groups = []
    with open(file_path, "r") as f:
        qid_prev = None
        count = 0
        for line in f:
            qid = line.split()[1].split(":")[1]
            if qid != qid_prev:
                if qid_prev is not None:
                    groups.append(count)
                count = 1
                qid_prev = qid
            else:
                count += 1
        groups.append(count)
    return groups


# Load a small sample dataset from sklearn
data = load_svmlight_file("./MQ2008/Fold5/train.txt")
group = load_group_info("./MQ2008/Fold5/train.txt")
X, y = data[0], data[1]
logging.info(f"X.shape {X.shape}")
logging.info(f"y.shape {y.shape}")
logging.info(f"sum(group) {sum(group)}")
logging.info(f"len(group) {len(group)}")


# Create LightGBM dataset
logging.info("Creating ORIGINAL LightGBM dataset :")
lgb_data = lgb.Dataset(X, label=y, group=group, free_raw_data=False)
logging.info(f"lgb_data.data.shape : {lgb_data.data.shape}")
logging.info(f"lgb_data.label.shape : {lgb_data.label.shape}")
logging.info(f"len(lgb_data.group) : {len(lgb_data.group)}")

logging.info("ORIGINAL dataset training ...")
PARAMS = {
    "objective": "lambdarank",
    "metric": "ndcg",
    "eval_at": 5,
    "num_iterations": 5,
    "random_state": 42,
}
lgb.train(PARAMS, lgb_data, valid_sets=[lgb_data], callbacks=[lgb.log_evaluation()])


# Save lgb Dataset
lgb_data.save_binary("lgb_data.bin")

# Load the dataset from the binary file
lgb_data_loaded = lgb.Dataset("lgb_data.bin", free_raw_data=False)
lgb_data_loaded.construct()


# We show that only feature_name & label is accessible
logging.info("\n\n\t\t\tData is not accessible from the saved binary :\n")
logging.info(f"lgb_data_loaded.feature_name) : {lgb_data_loaded.feature_name}")
logging.info(f"lgb_data_loaded.label) : {lgb_data_loaded.label}")
logging.info(f"lgb_data_loaded.label.shape) : {lgb_data_loaded.label.shape}")
logging.info(f"lgb_data_loaded.group) : {lgb_data_loaded.group}")
logging.info(f"lgb_data_loaded.data) : {lgb_data_loaded.data}")
logging.info(f"type(lgb_data_loaded.data)) : {type(lgb_data_loaded.data)}")
logging.info(f"lgb_data_loaded.get_data()) : {lgb_data_loaded.get_data()}")
logging.info(f"type(lgb_data_loaded.get_data())) : {type(lgb_data_loaded.get_data())}")
logging.info(f"lgb_data_loaded.num_data()) : {lgb_data_loaded.num_data()}")
logging.info(f"lgb_data_loaded.num_feature()) : {lgb_data_loaded.num_feature()}")


# Data is Here because it is able to train... Why can't we access it then ?
logging.info(
    "\n\n\t\t\tData is Here because it is able to train... Why can't we access it then ?\n"
)
lgb.train(
    PARAMS,
    lgb_data_loaded,
    valid_sets=[lgb_data_loaded],
    callbacks=[lgb.log_evaluation()],
)
```
Which outputs :

```sh
2024-07-01 13:41:43,996 - INFO : lgb.__version__ = 4.4.0
2024-07-01 13:41:44,145 - INFO : X.shape (9442, 46)
2024-07-01 13:41:44,146 - INFO : y.shape (9442,)
2024-07-01 13:41:44,146 - INFO : sum(group) 9442
2024-07-01 13:41:44,146 - INFO : len(group) 470
2024-07-01 13:41:44,146 - INFO : Creating ORIGINAL LightGBM dataset :
2024-07-01 13:41:44,146 - INFO : lgb_data.data.shape : (9442, 46)
2024-07-01 13:41:44,146 - INFO : lgb_data.label.shape : (9442,)
2024-07-01 13:41:44,146 - INFO : len(lgb_data.group) : 470
2024-07-01 13:41:44,146 - INFO : ORIGINAL dataset training ...
[LightGBM] [Info] Total Bins 9221
[LightGBM] [Info] Number of data points in the train set: 9442, number of used features: 40
[1]     training's ndcg@5: 0.766816
[2]     training's ndcg@5: 0.8209
[3]     training's ndcg@5: 0.838295
[4]     training's ndcg@5: 0.856369
[5]     training's ndcg@5: 0.865225
[LightGBM] [Info] Saving data to binary file lgb_data.bin
[LightGBM] [Info] Load from binary file lgb_data.bin
2024-07-01 13:41:44,725 - INFO : 

                        Data is not accessible from the saved binary :

2024-07-01 13:41:44,726 - INFO : lgb_data_loaded.feature_name) : ['Column_0', 'Column_1', 'Column_2', 'Column_3', 'Column_4', 'Column_5', 'Column_6', 'Column_7', 'Column_8', 'Column_9', 'Column_10', 'Column_11', 'Column_12', 'Column_13', 'Column_14', 'Column_15', 'Column_16', 'Column_17', 'Column_18', 'Column_19', 'Column_20', 'Column_21', 'Column_22', 'Column_23', 'Column_24', 'Column_25', 'Column_26', 'Column_27', 'Column_28', 'Column_29', 'Column_30', 'Column_31', 'Column_32', 'Column_33', 'Column_34', 'Column_35', 'Column_36', 'Column_37', 'Column_38', 'Column_39', 'Column_40', 'Column_41', 'Column_42', 'Column_43', 'Column_44', 'Column_45']
2024-07-01 13:41:44,726 - INFO : lgb_data_loaded.label) : [0. 0. 0. ... 0. 2. 0.]
2024-07-01 13:41:44,726 - INFO : lgb_data_loaded.label.shape) : (9442,)
2024-07-01 13:41:44,726 - INFO : lgb_data_loaded.group) : None
2024-07-01 13:41:44,726 - INFO : lgb_data_loaded.data) : lgb_data.bin
2024-07-01 13:41:44,726 - INFO : type(lgb_data_loaded.data)) : <class 'str'>
2024-07-01 13:41:44,726 - INFO : lgb_data_loaded.get_data()) : lgb_data.bin
2024-07-01 13:41:44,726 - INFO : type(lgb_data_loaded.get_data())) : <class 'str'>
2024-07-01 13:41:44,726 - INFO : lgb_data_loaded.num_data()) : 9442
2024-07-01 13:41:44,726 - INFO : lgb_data_loaded.num_feature()) : 46
2024-07-01 13:41:44,726 - INFO : 

                        Data is Here because it is able to train... Why can't we access it then ?

[LightGBM] [Info] Total Bins 9221
[LightGBM] [Info] Number of data points in the train set: 9442, number of used features: 40
[1]     training's ndcg@5: 0.766816
[2]     training's ndcg@5: 0.8209
[3]     training's ndcg@5: 0.838295
[4]     training's ndcg@5: 0.856369
[5]     training's ndcg@5: 0.865225
```

### PR DESCRIPTION
**Pull request by Plenitude-ai**
Fixes #6511

As discussed with James Lamb in https://github.com/microsoft/LightGBM/issues/6511#issuecomment-2235274432, a bit more documentation about the way the lgb.Dataset is actually constructed, with a particular representation of data, could be interesting for users.

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks very much for this!

As I mentioned in #6511, I'll push a commit here with the corresponding change for the R package, and then we can ask for other maintainers to review.

---

In the future when you use the pull-request-from-a-fork contributing process, I strongly recommend creating your own branch instead of using `main` / `master`. See https://github.com/microsoft/LightGBM/pull/6501#pullrequestreview-2132981614 for some reasons and https://github.com/scipy-conference/scipy2024_sprints_first_PR/blob/main/prs_and_reviewing.md for some advice on that.

**Comment by jameslamb** [comment]
I've also put "fixes #6511" in the description, so that will be automatically closed when we merge this. I believe this answers your final question there.

**Comment by jameslamb** [comment]
Thanks again @Plenitude-ai . I've pushed commits updating the formatting and adding the corresponding R changes. Please follow the instructions in https://github.com/microsoft/LightGBM/pull/6558#issuecomment-2237978344 (sorry, it's a Microsoft requirement), and one of the other maintainers here will review when they have time.

**Comment by Plenitude-ai** [comment]
Thanks again @jameslamb for your guidance. I realize it's a (very) modest one, but this is my first contribution to an open-source project so it's a great pleasure !
Thanks for your different advices on good practices in this process, i'll definitely follow them in the future.
@microsoft-github-policy-service agree company="Qwant"

**Comment by jameslamb** [comment]
@Plenitude-ai wonderful to hear it, thanks for choosing LightGBM for your first contribution! We really appreciate the help, and hope you'll consider contributing more in the future.

**Review by StrikerRUS — APPROVED** [review]
@Plenitude-ai Thanks a lot for your contribution!
And thanks @jameslamb for helping with R-package!
